Implement Required Members for VB

333fred opened the pull request:
Implements support for required members in VB. We support creating new instances of types with required members and ensuring that all required members are set in an object initializer. We do not support inheriting from a type with required members, as that would require emitting new metadata for these cases. Closes https://github.com/dotnet/roslyn/issues/61435.

Relates to test plan https://github.com/dotnet/roslyn/issues/57046

333fred commented:
@dotnet/roslyn-compiler @AlekseyTs this should be ready for review.

333fred commented:
@dotnet/roslyn-compiler @AlekseyTs for review please.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/VBResources.resx:
```diff
@@ -5668,4 +5668,16 @@
     <value>Analyzer reference specified multiple times</value>
   </data>
+  <data name="ERR_RequiredMemberMustBeSet" xml:space="preserve">
+    <value>Required member '{0}' must be set in the object initializer or attribute constructor.</value>
```
>or attribute constructor [](http://example.com/codeflow?start=71&length=24)

This sounds confusing. Perhaps it should say "in attribute arguments"?  
>Attribute  ::=
 [  AttributeModifier  :  ]  SimpleTypeName  [  OpenParenthesis  [  AttributeArguments  ]  CloseParenthesis  ]

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Binding/Binder_ObjectInitializer.vb:
```diff
@@ -508,6 +516,62 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+            creationSyntax As SyntaxNode,
+            diagnostics As BindingDiagnosticBag)
+
+            If constructor Is Nothing OrElse constructor.HasSetsRequiredMembers Then
```
>constructor Is Nothing [](http://example.com/codeflow?start=15&length=22)

In what situation this condition is true?

333fred replied:
Looks like this occurs when we see a struct with a `Protected` parameterless ctor, which we would not handle correctly here. I'll add a test and figure out what to do for the scenario.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Binding/Binder_ObjectInitializer.vb:
```diff
@@ -508,6 +516,62 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+                        Continue For
+                    End If
+
+                    If Not memberSymbol.Equals(requiredMember, TypeCompareKind.ConsiderEverything) Then
```
>TypeCompareKind.ConsiderEverything [](http://example.com/codeflow?start=63&length=34)

Should probably use ```AllIgnoreOptionsForVB``` here, unless we can demonstrate significance of them with a test.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Binding/Binder_ObjectInitializer.vb:
```diff
@@ -508,6 +516,62 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+                        DirectCast(TryCast(assignmentOperator.Left, BoundPropertyAccess)?.PropertySymbol, Symbol),
+                        TryCast(assignmentOperator.Left, BoundFieldAccess)?.FieldSymbol)
+
+                    If memberSymbol Is Nothing Then
```
>If memberSymbol Is Nothing Then [](http://example.com/codeflow?start=20&length=31)

Can we assert the receiver, since we are not checking it?

333fred replied:
I'm not certain what about the receiver you'd like me to assert?

AlekseyTs replied:
> I'm not certain what about the receiver you'd like me to assert?

Sorry, I meant an assert that the receiver is what we expect (placeholder?) of expected type.

333fred replied:
I guess I'm not sure what invariant you're hoping that this will verify.

333fred replied:
Nested member initializers on required members are an error in C# as well.

It appears that the receiver is not always present. Attributes do not pass one along, `ReceiverOpt` is nothing for that scenario.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/ConstraintsHelper.vb:
```diff
@@ -895,28 +894,34 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
-                            Else
-                                Return True
-                            End If
+                        If constraintError = ConstructorConstraintError.None AndAlso classType.IsMustInherit Then
```
>constraintError [](http://example.com/codeflow?start=27&length=15)

Consider adding case for None and a throw for an unexpected value in Case Else

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/ConstraintsHelper.vb:
```diff
@@ -895,28 +894,34 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
                     If diagnosticsBuilder IsNot Nothing Then
-                        ' "Type argument '{0}' must have a public parameterless instance constructor to satisfy the 'New' constraint for type parameter '{1}'."
-                        diagnosticsBuilder.Add(New TypeParameterDiagnosticInfo(typeParameter, ErrorFactory.ErrorInfo(ERRID.ERR_NoSuitableNewForNewConstraint2, typeArgument, typeParameter)))
+                        Select Case constraintError
```
>Select Case constraintError [](http://example.com/codeflow?start=24&length=27)

Consider adding case for None and a throw for an unexpected value in Case Else

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/ConstraintsHelper.vb:
```diff
@@ -895,28 +894,34 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+                    Dim constraintError = ConstructorConstraintError.NoPublicParameterlessConstructor
+
+                    If typeArgument.TypeKind = TypeKind.Class OrElse isStructure Then
                         Dim classType = DirectCast(typeArgument, NamedTypeSymbol)
```
>classType [](http://example.com/codeflow?start=28&length=9)

The name of the local is somewhat misleading now. Consider renaming.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/ConstraintsHelper.vb:
```diff
@@ -1086,28 +1091,55 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
-            Debug.Assert(type.TypeKind = TypeKind.Class)
+            Debug.Assert(type.TypeKind = TypeKind.Class OrElse type.TypeKind = TypeKind.Structure)
 
             Dim sourceNamedType = TryCast(type, SourceNamedTypeSymbol)
```
>Dim sourceNamedType = TryCast(type, SourceNamedTypeSymbol) [](http://example.com/codeflow?start=12&length=58)

This is a preexisting condition, but it looks like we are not dealing with constructed types properly here. Consider opening an issue to follow up.

333fred replied:
[https://github.com/dotnet/roslyn/issues/66564](https://github.com/dotnet/roslyn/issues/66564)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/ConstraintsHelper.vb:
```diff
@@ -1086,28 +1091,55 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             Dim sourceNamedType = TryCast(type, SourceNamedTypeSymbol)
 
-            If sourceNamedType IsNot Nothing AndAlso Not sourceNamedType.MembersHaveBeenCreated Then
+            If sourceNamedType IsNot Nothing AndAlso Not sourceNamedType.MembersHaveBeenCreated AndAlso sourceNamedType.TypeKind = TypeKind.Class Then
```
>sourceNamedType.TypeKind = TypeKind.Class [](http://example.com/codeflow?start=104&length=41)

Consider checking this first and not even calculating ```sourceNamedType``` otherwise.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/ConstraintsHelper.vb:
```diff
@@ -1086,28 +1091,55 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
                 ' we can end up here while we are building the set of members for the type.
                 ' Using InstanceConstructors property will send us into an infinite loop.
-                Return sourceNamedType.InferFromSyntaxIfClassWillHavePublicParameterlessConstructor()
+                Debug.Assert(sourceNamedType.TypeKind = TypeKind.Class)
```
>Debug.Assert(sourceNamedType.TypeKind = TypeKind.Class) [](http://example.com/codeflow?start=16&length=55)

The assert looks redundant, we just checked this condition explicitly.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/ConstraintsHelper.vb:
```diff
@@ -1086,28 +1091,55 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             Next
-            Return False
+
+            If Not synthesizedIfMissing Then
```
>If Not synthesizedIfMissing Then [](http://example.com/codeflow?start=12&length=32)

It is not obvious why is this needed. Is VB not synthesizing default constructor for structs?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/FieldSymbol.vb:
```diff
@@ -417,12 +419,6 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             End Get
         End Property
 
-        Private ReadOnly Property IFieldSymbol_IsRequired As Boolean Implements IFieldSymbol.IsRequired
```
>Private ReadOnly Property IFieldSymbol_IsRequired As Boolean Implements IFieldSymbol.IsRequired [](http://example.com/codeflow?start=8&length=95)

Please follow the pattern and keep this member. Adjust implementation instead.

333fred replied:
`IsReadOnly` in this type follows the same pattern I used.

AlekseyTs replied:
It is quite obvious that ```IsReadonly``` is an outlier and should be fixed too.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/PropertySymbol.vb:
```diff
@@ -571,12 +573,6 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             End Get
         End Property
 
-        Private ReadOnly Property IPropertySymbol_IsRequired As Boolean Implements IPropertySymbol.IsRequired
```
>Private ReadOnly Property IPropertySymbol_IsRequired As Boolean Implements IPropertySymbol.IsRequired [](http://example.com/codeflow?start=8&length=101)

Please follow the pattern and keep this member. Adjust implementation instead.

333fred replied:
I'm using the pattern used by this type. As examples, `IsReadOnly` and `IsWriteOnly` do the pattern I did here.

AlekseyTs replied:
See surrounding code for the pattern. It is quite possible that we have a couple of outliers (that should be fixed) and we shouldn't add more deviation. The original code did the right thing for these APIs. let's keep doing the right thing.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/SubstitutedNamedType.vb:
```diff
@@ -589,6 +590,18 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
 
+        Friend Overrides ReadOnly Property AllRequiredMembers As ImmutableSegmentedDictionary(Of String, Symbol)
+            Get
+                Return OriginalDefinition.AllRequiredMembers
```
>OriginalDefinition.AllRequiredMembers [](http://example.com/codeflow?start=23&length=37)

This doesn't look like the right thing to do. Shouldn't returned members be substituted?

AlekseyTs replied:
Including those from base, I guess

AlekseyTs replied:
BTW, what are we doing for C#?

333fred replied:
C# recalculates this on every type using `GetMembersUnordered`, so it should have the correct substitutions.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/UnboundGenericType.vb:
```diff
@@ -338,6 +339,18 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
 
+        Friend Overrides ReadOnly Property AllRequiredMembers As ImmutableSegmentedDictionary(Of String, Symbol)
+            Get
+                Return OriginalDefinition.AllRequiredMembers
```
>OriginalDefinition.AllRequiredMembers [](http://example.com/codeflow?start=23&length=37)

It looks like unbound generic types have only types, otherwise those should be substituted. What are we doing for C#?

333fred replied:
C# doesn't have a dedicate type for unbound generics, it uses ErrorTypeSymbols, which have no members and will calculate an empty set.

AlekseyTs replied:
Are we testing these code paths?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PEFieldSymbol.vb:
```diff
@@ -289,6 +290,16 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
+        Public Overrides ReadOnly Property IsRequired As Boolean
+            Get
+                If Not _lazyIsRequired.HasValue() Then
+                    _lazyIsRequired = PEModule.HasAttribute(Handle, AttributeDescription.RequiredMemberAttribute).ToThreeState()
```
>AttributeDescription.RequiredMemberAttribute [](http://example.com/codeflow?start=68&length=44)

Can this attribute be applied in VB source?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PEMethodSymbol.vb:
```diff
@@ -1288,6 +1316,24 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
+                    Return hasSetsRequiredMembersValue
+                End If
+
+                hasSetsRequiredMembersValue = DirectCast(ContainingModule, PEModuleSymbol).Module.HasAttribute(Handle, AttributeDescription.SetsRequiredMembersAttribute)
```
>SetsRequiredMembersAttribute [](http://example.com/codeflow?start=140&length=28)

Can this attribute be applied in source?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.vb:
```diff
@@ -1556,6 +1557,94 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
+
+        Private ReadOnly Property HasAnyDeclaredRequiredMembers As Boolean
+            Get
+                Return ContainingPEModule.Module.HasAttribute(Handle, AttributeDescription.RequiredMemberAttribute)
```
>RequiredMemberAttribute [](http://example.com/codeflow?start=91&length=23)

Can this attribute be applied in source?

333fred replied:
It can, but as with the other similar question, we aren't checking it. We're just not forbidding it.

AlekseyTs replied:
> It can, but as with the other similar question, we aren't checking it. We're just not forbidding it.

I think **not forbidding it** is going to complicate implementation for retargeting scenario. If one can apply attributes in source, they will be respected from metadata after we emit. Retargeting should "duplicate" this behavior, i.e. understand the attributes in source is if we emitted them to metadata and then imported them. I think it will be easier (and better) to forbid the attributes in source, unless we are willing to do the implementation and testing for retargeting.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.vb:
```diff
@@ -1556,6 +1557,94 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
+            End Get
+        End Property
+
+        Private Shared ReadOnly s_requiredMembersErrorSentinel As ImmutableSegmentedDictionary(Of String, Symbol) = ImmutableSegmentedDictionary(Of String, Symbol).Empty.Add("<error sentinel>", Nothing)
```
>Private Shared ReadOnly s_requiredMembersErrorSentinel As ImmutableSegmentedDictionary(Of String, Symbol) = ImmutableSegmentedDictionary(Of String, Symbol).Empty.Add("<error sentinel>", Nothing) [](http://example.com/codeflow?start=8&length=194)

Please declare all fields at the top of the type declaration

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.vb:
```diff
@@ -1556,6 +1557,94 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
+            Get
+                CalculateRequiredMembersIfRequired()
+                Debug.Assert(Not _lazyRequiredMembers.IsDefault)
+                Return _lazyRequiredMembers = s_requiredMembersErrorSentinel
```
>= [](http://example.com/codeflow?start=44&length=1)

```Is```?

333fred replied:
`ImmutableSegmentedDictionary` is a struct type (that wraps a reference type). `Is` does not work, but `=` checks the underlying dictionaries for equality.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.vb:
```diff
@@ -1556,6 +1557,94 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
+
+        Private Function TryCalculateRequiredMembers(ByRef requiredMembersBuilder As ImmutableSegmentedDictionary(Of String, Symbol).Builder) As Boolean
+            Dim lazyRequiredMembers = _lazyRequiredMembers
+            If lazyRequiredMembers = s_requiredMembersErrorSentinel Then
```
>If lazyRequiredMembers = s_requiredMembersErrorSentinel Then [](http://example.com/codeflow?start=12&length=60)

We checked ```_lazyRequiredMembers``` before calling this method. Why doing this again?

333fred replied:
This is a recursive function, base types should check this.

AlekseyTs replied:
> This is a recursive function, base types should check this.

I do not think we can rely on this recursion for bases, because base is not always a PENAmedTypeSymbol. We should request the information from AllRequiredMembers property instead.

333fred replied:
I was trying to avoid putting `_lazyRequiredMembers` on all `NamedTypeSymbol`s in VB, but I think you've convinced me I can't avoid it. I can certainly make it a bit smarter about caching the dictionary as it goes along.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.vb:
```diff
@@ -1556,6 +1557,94 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
+                Return False
+            End If
+
+            If TryCast(BaseTypeNoUseSiteDiagnostics, PENamedTypeSymbol)?.TryCalculateRequiredMembers(requiredMembersBuilder) = False Then
```
>TryCast(BaseTypeNoUseSiteDiagnostics, PENamedTypeSymbol) [](http://example.com/codeflow?start=15&length=56)

This is going to fail for generic constructed type.

333fred replied:
Added `GenericSubstitution_Inheritance_NoneSet`

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.vb:
```diff
@@ -1556,6 +1557,94 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
+                    Dim overriddenMember = TryCast(member, PropertySymbol)?.OverriddenProperty
+
+                    If (Not member.IsRequired()) OrElse
+                       member.Kind = SymbolKind.Field OrElse
```
>member.Kind = SymbolKind.Field OrElse [](http://example.com/codeflow?start=23&length=37)

Isn't the following check for ```overriddenMember``` sufficient?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.vb:
```diff
@@ -1556,6 +1557,94 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
+                    If (Not member.IsRequired()) OrElse
+                       member.Kind = SymbolKind.Field OrElse
+                       overriddenMember Is Nothing OrElse
+                       (Not overriddenMember.Equals(requiredMembersBuilder(member.Name), TypeCompareKind.ConsiderEverything)) Then
```
>ConsiderEverything [](http://example.com/codeflow?start=105&length=18)

Why ```ConsiderEverything```?

333fred replied:
Thinking about this a bit more, I actually don't think that I even need symbol comparison. The purpose of this check is "am I literally overriding this member", and reference equality should suffice for this.

AlekseyTs replied:
>  I literally overriding this member", and reference equality should suffice for this.

That is not a guarantee for generic scenarios for example. We might have equal, but distinct members from constructed base.

AlekseyTs replied:
BTW, since we are comparing for the purpose of overriding, we probably should pay attention to custom modifiers

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.vb:
```diff
@@ -1556,6 +1557,94 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
+                    If (Not member.IsRequired()) OrElse
+                       member.Kind = SymbolKind.Field OrElse
+                       overriddenMember Is Nothing OrElse
+                       (Not overriddenMember.Equals(requiredMembersBuilder(member.Name), TypeCompareKind.ConsiderEverything)) Then
```
>requiredMembersBuilder(member.Name) [](http://example.com/codeflow?start=52&length=35)

Consider merging with ```ContainsKey``` check (TryGetValue?).

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PEPropertySymbol.vb:
```diff
@@ -610,6 +612,16 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
+        Public Overrides ReadOnly Property IsRequired As Boolean
+            Get
+                If Not _lazyIsRequired.HasValue() Then
+                    _lazyIsRequired = _containingType.ContainingPEModule.Module.HasAttribute(Handle, AttributeDescription.RequiredMemberAttribute).ToThreeState()
```
>RequiredMemberAttribute [](http://example.com/codeflow?start=122&length=23)

Can this attribute be applied in VB source?

333fred replied:
It can, but I'm explicitly ignoring it. If, as a VB user, you want to apply the required member attribute to things yourself, you can, but you are responsible for getting it right and using it correctly within the same compilation.

AlekseyTs replied:
Then we have to support the attributes from source for retargeting.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Retargeting/RetargetingNamedTypeSymbol.vb:
```diff
@@ -546,5 +546,23 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Retargeting
+                    builder.Add(kvp.Key, RetargetingTranslator.Retarget(kvp.Value))
+                Next
+
+                Return builder.ToImmutable()
```
>Return builder.ToImmutable() [](http://example.com/codeflow?start=16&length=28)

I think all retargeting symbols are wrappers around symbols from VB module. So, they cannot have any required member "things". Instead of having "dead" code, let's assert that underlying symbols return false and empty sets form the new APIs and explicitly return false and empty sets.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/ConstraintsHelper.vb:
```diff
@@ -1086,28 +1091,55 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
         ''' Return true if the class type has a public parameterless constructor.
         ''' </summary>
-        Public Function HasPublicParameterlessConstructor(type As NamedTypeSymbol) As Boolean
+        Public Function HasPublicParameterlessConstructor(type As NamedTypeSymbol, synthesizedIfMissing As Boolean) As ConstructorConstraintError
```
>, synthesizedIfMissing As Boolean [](http://example.com/codeflow?start=81&length=33)

It looks like we can always infer this value from ```type.TypeKind = TypeKind.Structure```

333fred replied:
I was mainly shooting for matching the C# implementation, but we can probably simplify the VB impl here.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Tuples/TupleFieldSymbol.vb:
```diff
@@ -116,6 +116,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+
+        Public NotOverridable Overrides ReadOnly Property IsRequired As Boolean
+            Get
+                Return False
```
>Return False [](http://example.com/codeflow?start=16&length=12)

This doesn't look like a justified assumption. The ```_underlyingField``` is almost never a source field

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Tuples/TuplePropertySymbol.vb:
```diff
@@ -132,5 +132,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+
+        Public Overrides ReadOnly Property IsRequired As Boolean
+            Get
+                Return False
```
>Return False [](http://example.com/codeflow?start=16&length=12)

This doesn't look like a justified assumption. The ```_underlyingProperty``` is almost never a source property

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Wrapped/WrappedMethodSymbol.vb:
```diff
@@ -268,5 +268,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             Return Me.UnderlyingMethod.GetDocumentationCommentXml(preferredCulture, expandIncludes, cancellationToken)
         End Function
+
+        Friend NotOverridable Overrides ReadOnly Property HasSetsRequiredMembers As Boolean
```
>Friend NotOverridable Overrides ReadOnly Property HasSetsRequiredMembers As Boolean [](http://example.com/codeflow?start=8&length=83)

It doesn't feel appropriate to provide implementation at this level

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Wrapped/WrappedNamedTypeSymbol.vb:
```diff
@@ -213,5 +214,17 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             Return Me._underlyingType.GetGuidString(guidString)
         End Function
+
+        Friend Overrides ReadOnly Property AllRequiredMembers As ImmutableSegmentedDictionary(Of String, Symbol)
```
>Friend Overrides ReadOnly Property AllRequiredMembers As ImmutableSegmentedDictionary(Of String, Symbol) [](http://example.com/codeflow?start=8&length=104)

It doesn't feel appropriate to provide implementations at this level

AlekseyTs commented:
Done with review pass (commit 6), tests are not looked at

333fred commented:
@AlekseyTs addressed your feedback.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/SubstitutedNamedType.vb:
```diff
@@ -589,6 +590,29 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+                Dim members = GetMembersUnordered()
+                Dim builder = originalRequiredMembers.ToBuilder()
+                For Each kvp In originalRequiredMembers
+                    builder(kvp.Key) = SubstituteTypeParametersInMember(kvp.Value)
```
>SubstituteTypeParametersInMember(kvp.Value) [](http://example.com/codeflow?start=39&length=43)

Are we testing effect of this fix?

333fred replied:
`GenericSubstitution_NoneSet` and `GenericSubstitution_AllSet`.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/SubstitutedNamedType.vb:
```diff
@@ -589,6 +590,29 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+            Get
+                Dim originalRequiredMembers = OriginalDefinition.AllRequiredMembers
+                If originalRequiredMembers.IsEmpty OrElse originalRequiredMembers = s_requiredMembersErrorSentinel Then
+                    Return originalRequiredMembers
```
>Return originalRequiredMembers [](http://example.com/codeflow?start=20&length=30)

Do we test this code path for both conditions that can get us here?

333fred replied:
This will be one unified codepath for all symbols in the next revision, more similar to C#.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Tuples/TupleFieldSymbol.vb:
```diff
@@ -116,6 +116,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+
+        Public NotOverridable Overrides ReadOnly Property IsRequired As Boolean
+            Get
+                Return _underlyingField.IsRequired
```
>_underlyingField.IsRequired [](http://example.com/codeflow?start=23&length=27)

Are we testing effect of this fix?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Tuples/TupleMethodSymbol.vb:
```diff
@@ -146,5 +146,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+
+        Friend Overrides ReadOnly Property HasSetsRequiredMembers As Boolean
+            Get
+                Return _underlyingMethod.HasSetsRequiredMembers
```
>_underlyingMethod.HasSetsRequiredMembers [](http://example.com/codeflow?start=23&length=40)

Are we testing this code path?

333fred replied:
This specific codepath shouldn't be reachable, I'll add a throw.

AlekseyTs replied:
> This specific codepath shouldn't be reachable, I'll add a throw.

I think this assumption is not correct.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Tuples/TuplePropertySymbol.vb:
```diff
@@ -132,5 +132,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+
+        Public Overrides ReadOnly Property IsRequired As Boolean
+            Get
+                Return _underlyingProperty.IsRequired
```
> _underlyingProperty.IsRequired [](http://example.com/codeflow?start=22&length=31)

Are we testing effect of this fix?

333fred replied:
I'll add a direct test, I don't expect this to be directly observable.

AlekseyTs replied:
> I don't expect this to be directly observable except via the public API.

I think it can be observed by explicitly using an object creation expression rather than a tuple literal. At least it is observable in C# this way.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Tuples/TupleTypeSymbol.vb:
```diff
@@ -1125,5 +1125,16 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
 
+        Friend Overrides ReadOnly Property AllRequiredMembers As ImmutableSegmentedDictionary(Of String, Symbol)
+            Get
+                Return _underlyingType.AllRequiredMembers
```
>_underlyingType.AllRequiredMembers [](http://example.com/codeflow?start=23&length=34)

I do not think we can simply return members from the underlying type. Are we testing this code path?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Wrapped/WrappedNamedTypeSymbol.vb:
```diff
@@ -7,6 +7,7 @@ Imports System.Globalization
 Imports System.Runtime.InteropServices
 Imports System.Threading
 Imports Microsoft.Cci
+Imports Microsoft.CodeAnalysis.Collections
```
Consider reverting, this change is no longer necessary

333fred commented on src/Compilers/VisualBasic/Test/Semantic/Semantics/RequiredMembersTests.vb:
```diff
@@ -0,0 +1,821 @@
+            ' {
+            '     protected S() {}
+            '     public required int Prop { get; set; }
+            ' }
```
No, otherwise I would have just used C#. But VB doesn't even have the pseudo-syntax for me to succinctly describe what I'm creating.

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [1373082412](https://github.com/dotnet/roslyn/pull/66084#issuecomment-1373082412) [](http://example.com/codeflow?ancestors=1373082412,1373082412)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Lowering/StateMachineRewriter/SynthesizedContainer.vb:
```diff
@@ -321,5 +321,23 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+            End Get
+        End Property
+
+        Friend Overrides ReadOnly Property HasAnyDeclaredRequiredMembers As Boolean
```
>Overrides [](http://example.com/codeflow?start=15&length=9)

NotOverridable?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb:
```diff
@@ -1089,6 +1099,85 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+                If lazyRequiredMembers = s_requiredMembersErrorSentinel Then
+                    Return False
+                Else
+                    requiredMembersBuilder = lazyRequiredMembers.ToBuilder()
```
>lazyRequiredMembers.ToBuilder() [](http://example.com/codeflow?start=45&length=31)

Is this operation expensive? Allocates something? If so, perhaps we should consider postponing it until we actually need a builder.

333fred replied:
It does allocate a single instance, but it's fairly cheap. It holds onto the source immutable dictionary until mutation occurs.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb:
```diff
@@ -1089,6 +1099,85 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+
+        Friend MustOverride ReadOnly Property HasAnyDeclaredRequiredMembers As Boolean
+
+        Private Function TryCalculateRequiredMembersWithCache(ByRef requiredMembersBuilder As ImmutableSegmentedDictionary(Of String, Symbol).Builder) As Boolean
```
>ByRef requiredMembersBuilde [](http://example.com/codeflow?start=62&length=27)

```<Out>```?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb:
```diff
@@ -1089,6 +1099,85 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+            Return success
+        End Function
+
+        Private Function TryCalculateRequiredMembers(ByRef requiredMembersBuilder As ImmutableSegmentedDictionary(Of String, Symbol).Builder) As Boolean
```
>ByRef requiredMembersBuilder [](http://example.com/codeflow?start=53&length=28)

```<In> <Out>```?

333fred replied:
We seem to be inconsistent across the codebase (and the current file) on whether we apply these attributes. What effect do they actually have?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb:
```diff
@@ -1089,6 +1099,85 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+        End Function
+
+        Private Function TryCalculateRequiredMembers(ByRef requiredMembersBuilder As ImmutableSegmentedDictionary(Of String, Symbol).Builder) As Boolean
+            If BaseTypeNoUseSiteDiagnostics?.TryCalculateRequiredMembersWithCache(requiredMembersBuilder) = False Then
```
>TryCalculateRequiredMembersWithCache [](http://example.com/codeflow?start=45&length=36)

Wouldn't it be simpler to get ```AllRequiredMembers``` here and then party on it? Are we trying to reuse a builder across entire inheritance hierarchy? Is it really worth the complexity?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb:
```diff
@@ -1089,6 +1099,85 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+            For Each member In GetMembersUnordered()
+                Dim existingMember As Symbol = Nothing
+                ' Need to make sure that members from a base type weren't hidden by members from the current type. That is an error scenario
+                If requiredMembersBuilder.TryGetValue(member.Name, existingMember) Then
```
>requiredMembersBuilder.TryGetValue(member.Name, existingMember [](http://example.com/codeflow?start=19&length=62)

It looks like we are looking in a collection that is getting mutated, and strictly speaking, doesn't represent the set of required members from the base. It is quite possible that this isn't going to make a difference for conventional scenarios. But it is hard to tell. For example, VB understands overloading of properties. I can easily imagine when parameter-less property is required, and parameterized property next to it is not. I think it will be better  (more predictable, more robust, etc.) to do exactly what we intend to do, i.e. check against required members from base, from its ```AllRequiredMembers```. And this comment is probably applicable to C# implementation as well.

AlekseyTs replied:
For example, is VB going to be able to properly handle instantiation  of ```C2``` from this example?
[https://sharplab.io/#v2:EYLgtghglgdgNAFxFANgHwAICYAEBhARgFgAoAb1JypwCcBTARwFcp6ATHDAZh1gRwAKBHGQDmdBAG4AzhMkBfSopKls+XCHzFylagG0MBAHQAlJjARQwdI3gD2YAA6o6NAMquAblADGdaUYAkjBsdAAergByENYAgggINFDATAh0ABQwMXR2AGbpQgCUhQC6ulTcvBY4CAAWUNJ6fDhhJTgAvAB8LZKkyqq4hgDspBQk1JwALDgAKv4I6YXlIssTAPodODB0AO7qiyJCmwTyvePUyvJAA==](https://sharplab.io/#v2:EYLgtghglgdgNAFxFANgHwAICYAEBhARgFgAoAb1JypwCcBTARwFcp6ATHDAZh1gRwAKBHGQDmdBAG4AzhMkBfSopKls+XCHzFylagG0MBAHQAlJjARQwdI3gD2YAA6o6NAMquAblADGdaUYAkjBsdAAergByENYAgggINFDATAh0ABQwMXR2AGbpQgCUhQC6ulTcvBY4CAAWUNJ6fDhhJTgAvAB8LZKkyqq4hgDspBQk1JwALDgAKv4I6YXlIssTAPodODB0AO7qiyJCmwTyvePUyvJAA==)
```
class C1
{
    required public int P1 {get;set;}
 }

class C2 : C1
{
    [System.Runtime.CompilerServices.IndexerNameAttribute(nameof(P1))]
    public int this[int x] => x;
}

class Program
{
    void Test()
    {
        _ = new C2() {P1 = 1};
    }
}
```

Now imagine that some language declared both properties in the same type.

AlekseyTs replied:
Here is an example how VB can consume both properties by name at the same time:
[https://sharplab.io/#v2:EYLgtghglgdgNAGxAN2HALiKCA+AFAV2ASgGMACAYQQgGdaqBGAWAChyPyAlAUwgBMA8jAQBPcngBOAewAOPSenF5G5AIIMAkjHQ8A5grYBRGPyo16bNoWJlzdBpQBMbTuW0ALBVHSOW7N05XQI5BZAUEaQEGXgFhMQkZeUVlRgAKAA91LR19BQBKbPdcg0lgkIBxHnRykM5edAJJGHIABlq3EzMqmoCOLsS5BSUOoL7yAGUicgAVHlp0NPzRkIARKDByLIBeCVUAaj20xmXxgangY1N7eiA](https://sharplab.io/#v2:EYLgtghglgdgNAGxAN2HALiKCA+AFAV2ASgGMACAYQQgGdaqBGAWAChyPyAlAUwgBMA8jAQBPcngBOAewAOPSenF5G5AIIMAkjHQ8A5grYBRGPyo16bNoWJlzdBpQBMbTuW0ALBVHSOW7N05XQI5BZAUEaQEGXgFhMQkZeUVlRgAKAA91LR19BQBKbPdcg0lgkIBxHnRykM5edAJJGHIABlq3EzMqmoCOLsS5BSUOoL7yAGUicgAVHlp0NPzRkIARKDByLIBeCVUAaj20xmXxgangY1N7eiA)
```
Public Class C1
    ReadOnly Property P1 As Integer
End Class

Public Class C2
    Inherits C1
        
        Overloads ReadOnly Property P1(x As Integer) As Integer
        Get
            Return 0
        End Get
    End Property
        
    Sub Test()
            Dim x = P1 + P1(1)
    End Sub
End Class
```

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb:
```diff
@@ -1089,6 +1099,85 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
         ''' </summary>
         Friend MustOverride Function GetSynthesizedWithEventsOverrides() As IEnumerable(Of PropertySymbol)
 
+        Friend Overridable ReadOnly Property AllRequiredMembers As ImmutableSegmentedDictionary(Of String, Symbol)
```
>Overridable [](http://example.com/codeflow?start=15&length=11)

It looks like the default implementation doesn't use ```AllRequiredMembers``` and ```HasRequiredMembersError``` in the process of calculation, and, therefore, going to ignore the overridden behavior while traversing base types. I am going to mention once again that I think that we should not use recursion through ```TryCalculateRequiredMembersWithCache``` while traversing base types. We should use these properties instead.

333fred replied:
I'm resolving this by simply removing the overrides.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Symbol_Attributes.vb:
```diff
@@ -200,6 +200,8 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             ElseIf arguments.Attribute.IsTargetAttribute(Me, AttributeDescription.CompilerFeatureRequiredAttribute) Then
                 DirectCast(arguments.Diagnostics, BindingDiagnosticBag).Add(ERRID.ERR_DoNotUseCompilerFeatureRequired, arguments.AttributeSyntaxOpt.Location)
+            ElseIf arguments.Attribute.IsTargetAttribute(Me, AttributeDescription.RequiredMemberAttribute) Then
+                DirectCast(arguments.Diagnostics, BindingDiagnosticBag).Add(ERRID.ERR_DoNotUseRequiredMember, arguments.AttributeSyntaxOpt.Location)
```
>DirectCast(arguments.Diagnostics, BindingDiagnosticBag).Add(ERRID.ERR_DoNotUseRequiredMember, arguments.AttributeSyntaxOpt.Location) [](http://example.com/codeflow?start=16&length=132)

Consider documenting this breaking change

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb:
```diff
@@ -1089,6 +1099,85 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+                    Continue For
+                End If
+
+                requiredMembersBuilder(member.Name) = member
```
>requiredMembersBuilder(member.Name) = member [](http://example.com/codeflow?start=16&length=44)

It feels like we should not be doing this when ```HasAnyDeclaredRequiredMembers``` is false.

333fred replied:
We'll never get here if `HasAnyDeclaredRequiredMembers` is false.

333fred replied:
I'll add an assert.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb:
```diff
@@ -1089,6 +1099,85 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+
+                    If (Not member.IsRequired()) OrElse
+                       overriddenMember Is Nothing OrElse
+                       overriddenMember IsNot existingMember Then
```
>overriddenMember IsNot existingMember [](http://example.com/codeflow?start=23&length=37)

This might fail to match members in a generic context for example.

333fred replied:
Looks like this is correct, added `GenericSubstitution_InheritanceAndOverride_NoneSet`

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb:
```diff
@@ -4042,6 +4043,24 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+            End Get
+        End Property
+
+        Friend Overrides ReadOnly Property HasAnyDeclaredRequiredMembers As Boolean
```
>Overrides [](http://example.com/codeflow?start=15&length=9)

NotOverridable?

AlekseyTs commented:
Done with review pass (commit 8), tests are not looked at. Also, it looks like a bunch of CI jobs is failing

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:
```diff
@@ -2442,6 +2458,88 @@ public class Derived : Base
+        // Equivalent to 
+        // class Base
+        // {
+        //     public required modopt(object) int Prop1 { get; set; }
```
>public required modopt(object) int Prop1 { get; set; } [](http://example.com/codeflow?start=15&length=54)

virtual?

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/Semantics/RequiredMembersTests.vb:
```diff
@@ -0,0 +1,1283 @@
+<RequiredMember>
+ ~~~~~~~~~~~~~~
+BC37325: 'System.Runtime.CompilerServices.RequiredMemberAttribute' is reserved for compiler usage only.
+    <RequiredMember>
```
><RequiredMember> [](http://example.com/codeflow?start=4&length=16)

Consider adding a comment in source to distinguish this location from the one below.

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/Semantics/RequiredMembersTests.vb:
```diff
@@ -0,0 +1,1283 @@
+            Dim comp = CreateCompilation("
+Class C
+    Sub Main()
+        Dim t = (1, 2)
```
>Dim t = (1, 2) [](http://example.com/codeflow?start=8&length=14)

Why isn't this an error? It looks like ```AnotherField``` and ```Property``` are not set.

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/Semantics/RequiredMembersTests.vb:
```diff
@@ -0,0 +1,1283 @@
+            Dim comp = CreateCompilation("
+Class C
+    Sub Main()
+        Dim t = (1, 2)
```
> (1, 2) [](http://example.com/codeflow?start=15&length=7)

Consider also testing scenario with ```new System.ValueTuple(of integer, integer)(1,2)```

AlekseyTs replied:
And ```new System.ValueTuple(of integer, integer)()``` 

BTW, do we test scenarios like that for C#?

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/Semantics/RequiredMembersTests.vb:
```diff
@@ -0,0 +1,1283 @@
+        public required int AnotherField;
+        public required int Property { get; set; }
+
+        public ValueTuple(T1 item1, T2 item2)
```
>public ValueTuple(T1 item1, T2 item2) [](http://example.com/codeflow?start=8&length=37)

I think we should also test scenario with SetsRequiredMembers

333fred replied:
Talked offline. This will be entirely blocked (in C# as well).

333fred commented on src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb:
```diff
@@ -1089,6 +1099,85 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+                    Continue For
+                End If
+
+                requiredMembersBuilder(member.Name) = member
```
I'll add a test with a doubly-overridden example, but it does not cause an issue.

AlekseyTs replied:
> I'll add a test with a doubly-overridden example, but it does not cause an issue.

This might simply be due to the fact that the test doesn't test the "right" scenario. Is there an explanation why this behavior won't cause an issue?

333fred replied:
The code that was in question is gone now.

AlekseyTs replied:
> I'll add a test with a doubly-overridden example, but it does not cause an issue.

Could you point me to the test? Also, was the test failing on commit 8?

333fred replied:
I've added them to all the tests like `EnforcedRequiredMembers_Override_NoneSet_01`. I'm unsure whether the test was failing on iteration 8.

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [1398786376](https://github.com/dotnet/roslyn/pull/66084#issuecomment-1398786376) [](http://example.com/codeflow?ancestors=1398786376,1398786376)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/VBResources.resx:
```diff
@@ -5668,4 +5668,19 @@
+  <data name="ERR_NewConstraintCannotHaveRequiredMembers" xml:space="preserve">
+    <value>'{2}' cannot satisfy the 'New' constraint on parameter '{1}' in the generic type or or method '{0}' because '{2}' has required members.</value>
+  </data>
+  <data name="ERR_DoNotUseRequiredMember" xml:space="preserve">
```
>ERR_DoNotUseRequiredMember [](http://example.com/codeflow?start=14&length=26)

It looks like there are some test failures around this error code

AlekseyTs commented:
>                 Return ImmutableSegmentedDictionary(Of String, Symbol).Empty

I assume this wasn't correct implementation either. A source type can inherit from a metadata type with required members. A we testing scenarios like this?

---
Refers to: src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb:4049 in 9ae7f8d. [](commit_id = 9ae7f8dc4cad62cd410a7eb6d280162eda6da017, deletion_comment = True)

333fred commented:
>                 Return ImmutableSegmentedDictionary(Of String, Symbol).Empty

`CannotInheritFromTypesWithRequiredMembers`. I'll add some examples of trying to instanciate such types.

---
In reply to: [1406668499](https://github.com/dotnet/roslyn/pull/66084#issuecomment-1406668499) [](http://example.com/codeflow?ancestors=1406668499,1406668499)

---
Refers to: src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb:4049 in 9ae7f8d. [](commit_id = 9ae7f8dc4cad62cd410a7eb6d280162eda6da017, deletion_comment = True)

AlekseyTs commented:
/azp run

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Binding/Binder_Conversions.vb:
```diff
@@ -1644,6 +1644,11 @@ DoneWithDiagnostics:
                 ' Come back to this, what about locations? (https:'github.com/dotnet/roslyn/issues/11013)
                 targetType = destTupleType.WithElementNames(sourceTuple.ArgumentNamesOpt)
+
+                If Not destTupleType.AllRequiredMembers.IsEmpty Then
```
>If Not destTupleType.AllRequiredMembers.IsEmpty Then [](http://example.com/codeflow?start=16&length=52)

The approach of intercepting tuple types during binding feels very fragile. I think it would be better to mark the type symbol as bad via ```GetUseSiteInfo``` API. While we are calculating the diagnostics, we probably can check if the type has the "tuple" name and arity between 1 and 8, and add ```ERR_UnsupportedType1`` if it has required members. We are not calculating use-site errors for types declared in source, but that shouldn't be a problem because VB cannot declare a type with required members in source.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Binding/Binder_ObjectInitializer.vb:
```diff
@@ -508,6 +517,73 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+                Return
+            End If
+
+            If TypeOf containingType Is SourceMemberContainerTypeSymbol Then
```
>TypeOf containingType Is SourceMemberContainerTypeSymbol [](http://example.com/codeflow?start=15&length=56)

This condition won't work for constructed types and for retargeting scenario as well. I do no think suppression of the diagnostics is worth added complexity. I would get rid of this ```If```

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/ConstraintsHelper.vb:
```diff
@@ -1110,28 +1094,61 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+
         ''' <summary>
         ''' Return true if the class type has a public parameterless constructor.
         ''' </summary>
```
>constructor.ContainingType [](http://example.com/codeflow?start=34&length=26)

```type```?

333fred replied:
It looks like codeflow messed this comment up, I'm not sure what it was supposed to be on, but I see it on the Function header of `HasPublicParameterlessConstructor`

AlekseyTs replied:
> It looks like codeflow messed this comment up, I'm not sure what it was supposed to be on

I think that comment was about ```constructor.ContainingType``` in 
```TypeOf constructor.ContainingType Is SourceMemberContainerTypeSymbol``` condition below. The condition is gone now.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/ConstraintsHelper.vb:
```diff
@@ -1110,28 +1094,61 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+
         ''' <summary>
         ''' Return true if the class type has a public parameterless constructor.
         ''' </summary>
```
>If TypeOf constructor.ContainingType Is SourceMemberContainerTypeSymbol Then [](http://example.com/codeflow?start=24&length=76)

I think this check won't work for retargeting scenarios. I do not think it is worth adding extra complexity with the goal to suppress extra errors. If we really think it is important to do, we should consider changing implementation of ```AllRequiredMembers``` to return an empty set instead. And, perhaps, adjust behavior of ```HasRequiredMembersError``` as well.

333fred replied:
It looks like codeflow messed this comment up, I'm not sure what it was supposed to be on, but I see it on the Function header of `HasPublicParameterlessConstructor`

AlekseyTs replied:
> It looks like codeflow messed this comment up, I'm not sure what it was supposed to be on ...

I think the comment was about ```If TypeOf constructor.ContainingType Is SourceMemberContainerTypeSymbol``` added below.

333fred replied:
The code in question has been removed.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb:
```diff
@@ -1089,6 +1099,102 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+            Get
+                TryCalculateRequiredMembersWithCache(Nothing)
+                Debug.Assert(Not _lazyRequiredMembers.IsDefault)
+                Return _lazyRequiredMembers
```
>_lazyRequiredMembers [](http://example.com/codeflow?start=23&length=20)

It looks like here we can return value of ```s_requiredMembersErrorSentinel```. Is this intentional? This feels like an exposure of an implementation detail, which might really confuse some consumers because content of ```s_requiredMembersErrorSentinel``` is "malformed". Would it be better to return an empty dictionary instead of ```s_requiredMembersErrorSentinel```?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb:
```diff
@@ -1089,6 +1099,102 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+
+                    If (Not member.IsRequired()) OrElse
+                       overriddenMember Is Nothing OrElse
+                       overriddenMember.Equals(member, TypeCompareKind.IgnoreTupleNames) Then
```
>overriddenMember.Equals(member, TypeCompareKind.IgnoreTupleNames) [](http://example.com/codeflow?start=23&length=65)

It looks like we lost negation on this condition, but all tests keep passing. Is there a test gap?

333fred replied:
Added `EnforcedRequiredMembers_OverriddenFromMetadata`.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb:
```diff
@@ -1089,6 +1099,102 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+            Return success
+        End Function
+
+        Private Function TryCalculateRequiredMembers(<[In], Out> ByRef requiredMembersBuilder As ImmutableSegmentedDictionary(Of String, Symbol).Builder) As Boolean
```
><[In], Out> ByRef requiredMembersBuilder As ImmutableSegmentedDictionary(Of String, Symbol).Builder [](http://example.com/codeflow?start=53&length=99)

I think that "dragging" the builder along add a fair amount  of complexity. Is this a case of premature optimization, or was this added to address a real observable problem?

333fred replied:
I don't feel that it adds a significant amount of complexity. The base calls will still need to occur, we'll still need to have the dual builder of members for the current type and dictionary of members in the base type that we have below. It would remove a parameter from this method and replace a call to `TryCalculateRequiredMembers` with `HasRequiredMembersError`, but the logic would not be significantly affected.

AlekseyTs replied:
I am not talking about computational complexity, but the complexity of the logic. In my opinion the current logic is more complex than it could be. Unnecessary complexity is a source of bugs. So far I haven't heard what advantage do we get for taking this risk.

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/Semantics/RequiredMembersTests.vb:
```diff
@@ -0,0 +1,1842 @@
+        End Sub
+
+        <Fact>
+        Public Sub GenericSubstitution_Unbound()
```
>Unbound [](http://example.com/codeflow?start=39&length=7)

Does this test really cover ```UnboundGenericType``` symbol? Probably not, because ```UnboundGenericType``` shouldn't have any members.

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/Semantics/RequiredMembersTests.vb:
```diff
@@ -0,0 +1,1842 @@
+        Dim t1 = (1, 2)
+        Dim t2 As (Integer, Integer) = (3, Nothing)
+        Dim t3 = New (Integer, Integer)(4, 5)
+        Dim t4 = New System.ValueTuple(Of Integer, Integer)(6, 7)
```
>Dim t4 = New System.ValueTuple(Of Integer, Integer)(6, 7) [](http://example.com/codeflow?start=8&length=57)

I think this should be an error as a simple tuple literal is an error. The types are supposed to be unified and supposed to be handled uniformly.

333fred replied:
After online discussion, I've removed the checking from binding and added it to lowering of tuple literal expressions.

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/Semantics/RequiredMembersTests.vb:
```diff
@@ -0,0 +1,1842 @@
+            Dim comp = CreateCompilation("
+Module M
+    Sub Main()
+        Dim c = New C2()
```
>Dim c = New C2() [](http://example.com/codeflow?start=8&length=16)

I think it would be good to test success scenario as well, i.e. with initializer for P1.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Binding/Binder_ObjectInitializer.vb:
```diff
@@ -508,6 +517,73 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+                        Continue For
+                    End If
+
+                    If Not memberSymbol.Equals(requiredMember, TypeCompareKind.AllIgnoreOptionsForVB) Then
```
>If Not memberSymbol.Equals(requiredMember, TypeCompareKind.AllIgnoreOptionsForVB) Then [](http://example.com/codeflow?start=20&length=86)

Are we testing scenarios when ```memberSymbol``` is an override?

333fred replied:
`EnforcedRequiredMembers_Inheritance_*`

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/Semantics/RequiredMembersTests.vb:
```diff
@@ -0,0 +1,1842 @@
+        Dim t1 = (1, 2)
+        Dim t2 As (Integer, Integer) = (3, Nothing)
+        Dim t3 = New (Integer, Integer)(4, 5)
+        Dim t4 = new System.ValueTuple(Of Integer, Integer)(6, 7)
```
>Dim t4 = new System.ValueTuple(Of Integer, Integer)(6, 7) [](http://example.com/codeflow?start=8&length=57)

Consider also testing ```new System.ValueTuple(of integer, integer)()```

AlekseyTs replied:
Plus ```Dim t2 As (Integer, Integer) = Nothing``` and ```Dim t2 As System.ValueTuple(of integer, integer) = Nothing```

AlekseyTs replied:
BTW, it looks like the following is a valid C# code. Is this expected?
```
S1 x = default;

struct S1
{
    public required int F;
}
```

333fred replied:
> BTW, it looks like the following is a valid C# code. Is this expected?

Yes. I forgot to explicitly document that in the spec, but it is expected. I'll do that.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/ErrorTypeSymbol.vb:
```diff
@@ -7,6 +7,7 @@ Imports System.Collections.Immutable
 Imports System.Runtime.InteropServices
 Imports System.Text
 Imports System.Threading
+Imports Microsoft.CodeAnalysis.Collections
```
Is this Imports necessary?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/SubstitutedNamedType.vb:
```diff
@@ -7,6 +7,7 @@ Imports System.Collections.Immutable
 Imports System.Globalization
 Imports System.Runtime.InteropServices
 Imports System.Threading
+Imports Microsoft.CodeAnalysis.Collections
```
Is this Imports necessary?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/UnboundGenericType.vb:
```diff
@@ -9,6 +9,7 @@ Imports System.Globalization
 Imports System.Runtime.InteropServices
 Imports System.Text
 Imports System.Threading
+Imports Microsoft.CodeAnalysis.Collections
```
Is this Imports necessary?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/AnonymousTypes/PublicSymbols/AnonymousTypeOrDelegatePublicSymbol.vb:
```diff
@@ -5,6 +5,7 @@
 Imports System.Collections.Immutable
 Imports System.Runtime.InteropServices
 Imports System.Threading
+Imports Microsoft.CodeAnalysis.Collections
```
Is this Imports necessary?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/AnonymousTypes/SynthesizedSymbols/AnonymousTypeOrDelegateTemplateSymbol.vb:
```diff
@@ -7,6 +7,7 @@ Imports System.Runtime.InteropServices
 Imports System.Threading
 Imports Microsoft.Cci
 Imports Microsoft.CodeAnalysis
+Imports Microsoft.CodeAnalysis.Collections
```
Is this Imports necessary?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Source/SourceNamedTypeSymbol_ComClass.vb:
```diff
@@ -4,6 +4,7 @@
 
 Imports System.Collections.Immutable
 Imports System.Runtime.InteropServices
+Imports Microsoft.CodeAnalysis.Collections
```
Is this Imports necessary?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/SynthesizedSymbols/SynthesizedEventDelegateSymbol.vb:
```diff
@@ -5,6 +5,7 @@
 Imports System.Collections.Immutable
 Imports System.Runtime.InteropServices
 Imports System.Threading
+Imports Microsoft.CodeAnalysis.Collections
```
Is this Imports necessary?

AlekseyTs commented on src/ExpressionEvaluator/VisualBasic/Source/ExpressionCompiler/Symbols/EENamedTypeSymbol.vb:
```diff
@@ -6,6 +6,7 @@ Imports System.Collections.Immutable
 Imports System.Runtime.InteropServices
 Imports System.Threading
 Imports Microsoft.Cci
+Imports Microsoft.CodeAnalysis.Collections
```
Is this Imports necessary?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/ConstraintsHelper.vb:
```diff
@@ -1110,28 +1118,61 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+                End If
             End If
 
+            Dim hasRequiredMembers = type.AllRequiredMembers.Count > 0
```
>AllRequiredMembers [](http://example.com/codeflow?start=42&length=18)

Can we get an error sentinel here? Are we testing a scenario like that?

333fred replied:
`GenericConstrainedToNew_Forbidden_HasRequiredMembersError`

AlekseyTs commented:
Done with review pass (commit 12)

333fred commented on src/Compilers/VisualBasic/Test/Semantic/Semantics/RequiredMembersTests.vb:
```diff
@@ -0,0 +1,1842 @@
+        Dim t3 = New (Integer, Integer)(4, 5)
+                     ~~~~~~~~~~~~~~~~~~
+BC37321: Required member 'Public AnotherField As Integer' must be set in the object initializer or attribute arguments.
+        Dim t4 = new System.ValueTuple(Of Integer, Integer)(6, 7)
```
From offline discussion, we won't have an unsupported type error here.

333fred commented on src/Compilers/VisualBasic/Test/Semantic/Semantics/RequiredMembersTests.vb:
```diff
@@ -0,0 +1,1842 @@
+        Dim t3 = New (Integer, Integer)(4, 5)
+                     ~~~~~~~~~~~~~~~~~~
+BC37321: Required member 'Public AnotherField As Integer' must be set in the object initializer or attribute arguments.
+        Dim t4 = new System.ValueTuple(Of Integer, Integer)(6, 7)
```
For the `New` case, it'll work. I don't think there's anything special to test here, this is just a standard object creation, which is plenty covered by other tests in the file.

AlekseyTs replied:
> For the ```New``` case, it'll work. I don't think there's anything special to test here, this is just a standard object creation, which is plenty covered by other tests in the file.

I think tuple types are special enough to explicitly test the success scenario for them too.

AlekseyTs replied:
The thread was marked as resolved, but I didn't actually see a test for the scenario

333fred commented on src/Compilers/VisualBasic/Portable/Symbols/Source/SourceNamedTypeSymbol.vb:
```diff
@@ -1404,6 +1404,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             If declaredOrDefaultBase IsNot Nothing Then
                 ReportUseSiteInfoForBaseType(declaredOrDefaultBase, declaredBase, diagnostics)
+
+                If declaredOrDefaultBase.AllRequiredMembers.Count > 0 Then
```
Adding `CannotInheritFromTypesWithRequiredMembers_WithMalformedRequiredMemberList`.

AlekseyTs commented:
@333fred It looks like there are legitimate test failures

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Binding/Binder_Conversions.vb:
```diff
@@ -1042,6 +1042,15 @@ DoneWithDiagnostics:
 
+            Dim targetNamedType = TryCast(targetType, NamedTypeSymbol)
+            If argument.IsNothingLiteral() AndAlso targetNamedType IsNot Nothing AndAlso targetType.IsStructureType() Then
+                ' Check to see if we're creating a default value of a structure with required members
```
>' Check to see if we're creating a default value of a structure with required members [](http://example.com/codeflow?start=16&length=85)

I am not sure why are we checking this. It looks like C# doesn't care for similar scenario:
```
class C1
{
    S1 Test()
    {
        return default;
    }
}

struct S1
{
    required public int F;
}
```

333fred replied:
I was responding to your comment on a previous iteration, under the assumption that we would change the C# code. However, since we've since decided to spec C# as not requiring members for structs on `default`, I'll be reverting this part of the change and updating the tests.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Binding/Binder_Conversions.vb:
```diff
@@ -1042,6 +1042,15 @@ DoneWithDiagnostics:
+            If argument.IsNothingLiteral() AndAlso targetNamedType IsNot Nothing AndAlso targetType.IsStructureType() Then
+                ' Check to see if we're creating a default value of a structure with required members
+                If targetNamedType.AllRequiredMembers.Count > 0 OrElse targetNamedType.HasRequiredMembersError Then
+                    Dim structCtor As MethodSymbol = targetNamedType.InstanceConstructors.FirstOrDefault(Function(ctor) ctor.ParameterCount = 0)
```
>targetNamedType.InstanceConstructors.FirstOrDefault(Function(ctor) ctor.ParameterCount = 0) [](http://example.com/codeflow?start=53&length=91)

Is conversion going to use any constructor?

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/Semantics/RequiredMembersTests.vb:
```diff
@@ -0,0 +1,2451 @@
+
+            Dim comp = CreateCompilation(vbCode, {cComp.EmitToImageReference()})
+            comp.AssertTheseDiagnostics(<expected>
+BC37321: Required member 'Public F As Integer' must be set in the object initializer or attribute arguments.
```
It doesn't look like we produce an error for a similar scenario in C# ([https://sharplab.io/#v2:EYLgtghglgdgNAFxFANnAJiA1AHwAIBMABAMICMAsAFADe1RDReZAbEwCxECy0MAFAEp6jOlUbiiAZSIBnIgF4i6AKYAzCAFcUCANzCGAX2pGq1angDMshACcNAYwRTqo8ZaI3lARw1RP6IlgnADE9KgMgA=](https://sharplab.io/#v2:EYLgtghglgdgNAFxFANnAJiA1AHwAIBMABAMICMAsAFADe1RDReZAbEwCxECy0MAFAEp6jOlUbiiAZSIBnIgF4i6AKYAzCAFcUCANzCGAX2pGq1angDMshACcNAYwRTqo8ZaI3lARw1RP6IlgnADE9KgMgA=)):
```
class C1
{
    static void Main()
    {
        S s = default;
    }
}


public struct S
{
    public required int F;
}
```

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/Semantics/RequiredMembersTests.vb:
```diff
@@ -0,0 +1,2451 @@
+
+            Dim comp = CreateCompilation(vbCode, {cComp.EmitToImageReference()})
+            comp.AssertTheseDiagnostics(<expected>
+BC37321: Required member 'Public F As Integer' must be set in the object initializer or attribute arguments.
```
It doesn't look like we produce an error for a similar scenario in C#

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/Semantics/RequiredMembersTests.vb:
```diff
@@ -0,0 +1,2451 @@
+Module M
+    Sub Main()
+        Dim d1 = New Derived()
+        'Dim d2 = New Derived(1)
```
>'Dim d2 = New Derived(1) [](http://example.com/codeflow?start=8&length=24)

Is commenting out this line intentional?

AlekseyTs commented:
Done with review pass (commit 13)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [1470886843](https://github.com/dotnet/roslyn/pull/66084#issuecomment-1470886843) [](http://example.com/codeflow?ancestors=1470886843,1470886843)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_TupleLiteralExpression.vb:
```diff
@@ -78,6 +81,8 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                                     hasErrors:=True)
                     End If
 
+                    Binder.CheckRequiredMembersInObjectInitializer(tuple8Ctor, tuple8Ctor.ContainingType, initializers:=ImmutableArray(Of BoundExpression).Empty, syntax, _diagnostics)
```
>Binder.CheckRequiredMembersInObjectInitializer(tuple8Ctor, tuple8Ctor.ContainingType, initializers:=ImmutableArray(Of BoundExpression).Empty, syntax, _diagnostics) [](http://example.com/codeflow?start=20&length=163)

Do we have a test where this line of code makes an observable difference?

333fred replied:
Yes, `TupleWithRequiredFields` has a tuple8Ctor with a required member.

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/Semantics/RequiredMembersTests.vb:
```diff
@@ -0,0 +1,2435 @@
+            Dim csharpReference = CreateCSharpCompilationWithRequiredMembers(csharp).EmitToImageReference()
+
+            Dim vb = CreateCompilation("
+Class VbDerivedBase
```
>VbDerivedBase [](http://example.com/codeflow?start=6&length=13)

I think we should verify what all the new APIs return for this type.

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/Semantics/RequiredMembersTests.vb:
```diff
@@ -0,0 +1,2435 @@
+    Inherits Base
+End Class
+
+Class VbDerivedDerived
```
>VbDerivedDerived [](http://example.com/codeflow?start=6&length=16)

And for this type as well

AlekseyTs commented:
Done with review pass (commit 17)

AlekseyTs reviewed (approved):
LGTM (commit 19)

jcouv commented on src/Compilers/VisualBasic/Portable/Symbols/Retargeting/RetargetingFieldSymbol.vb:
```diff
@@ -206,6 +206,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Retargeting
 
+        Public Overrides ReadOnly Property IsRequired As Boolean
+            Get
+                Return _underlyingField.IsRequired
```
Would Aleksey's comment in `RetargetingNamedTypeSymbol` apply here as well? If this is always false, consider asserting so. Also applies to other retargeting symbols.

jcouv replied:
For clarification, the comment I was referring to was:
> I think all retargeting symbols are wrappers around symbols from VB module. So, they cannot have any required member "things". Instead of having "dead" code, let's assert that underlying symbols return false and empty sets form the new APIs and explicitly return false and empty sets.   

The question here is, can we add an assertion that `_underlyingField.IsRequired` is False?

333fred replied:
Tightened the asserts and added a test that adds a required member to a base type that requires retargeting a VB symbol to include required members.

jcouv commented on docs/compilers/Visual Basic/Compiler Breaking Changes - DotNet 7.md:
```diff
@@ -0,0 +1,8 @@
+# This document lists known breaking changes in Roslyn after .NET 6 all the way to .NET 7.
+
+## `RequiredMembersAttribute` cannot be manually applied
```
All the errors that were introduced are breaking changes: types with required members no longer fulfill the `New()` constraint and cannot be inherited, etc

333fred replied:
There was already next to no chance you could inherit from a type with required members in VB: only constructors with `SetsRequiredMembersAttribute` could be called, as VB reports errors for using a constructor without it.

jcouv commented on src/Compilers/VisualBasic/Test/Semantic/Semantics/RequiredMembersTests.vb:
```diff
@@ -0,0 +1,2449 @@
+            Dim csharpComp = CreateCSharpCompilation("
+namespace System
+{
+    public struct ValueTuple<T1, T2>
```
I don't love scenarios where system types are defined in unconventional ways, but looking at where we do `New BoundObjectCreationExpression` in the codebase I also noticed a few related to `System.Nullable`. 
I would rather assume that `Nullable` (and `ValueTuple`) are defined properly than bother with handling those scenarios, but just want to point it out for completeness.

jcouv commented on src/Compilers/VisualBasic/Test/Semantic/Semantics/RequiredMembersTests.vb:
```diff
@@ -0,0 +1,2449 @@
+Imports Microsoft.CodeAnalysis.Test.Utilities
+Imports Microsoft.CodeAnalysis.VisualBasic.Symbols
+Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
+Imports Moq
```
Is Moq needed?

jcouv commented on src/Compilers/VisualBasic/Test/Semantic/Semantics/RequiredMembersTests.vb:
```diff
@@ -0,0 +1,2449 @@
+            Dim prop = c.GetMember(Of PropertySymbol)("Prop")
+            Dim field = c.GetMember(Of FieldSymbol)("Field")
+
+            AssertEx.Equal(Of Symbol)({field, prop}, From kvp In c.AllRequiredMembers
```
nit: For tests that verify new symbol APIs like `AllRequiredMembers`, it would also be good to make the type of the symbol apparent in the test (for example, `RetargetingPropertySymbol`, ...). 
That makes it easy to verify which symbols were covered

jcouv reviewed (commented):
Done with review pass (iteration 19)

333fred commented on src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:
```diff
@@ -2508,28 +2606,118 @@ public class Derived : Base
+    /// }
+    /// </code>
+    /// </summary>
+    private const string ShadowingBaseAndDerivedIL = @"
```
It's not legal C# code, we can't get metadata for it. That's why I was originally using VB, because that was easier to write than IL.

333fred replied:
It's not legal C# code, we can't get metadata for it. That's why I was originally using VB, because that was easier to write than IL.

AlekseyTs reviewed (approved):
LGTM (commit 22)
